### ISSUE TITLE
meta_update function

### ISSUE THREAD

**Issue by WangWenshan** [katerakelly/pytorch-maml#8]
Hi Kate Rakelly,
Thank you for sharing the MAML source code. 
Here I have a few questions: 
1  Since you have obtained the meta_grads in the function “forward(*)”, why not directly utilize them to update the parameters of self.net. (param - learning rate*gradient)  (Here, it seems that you want to achieve that by defining a “meta_update(*)” function. And you make the annotation: “We use a dummy forward / backward pass to get the correct grads into self.net”. for this command: “loss, out = forward_pass(self.net, in_, target)”
I’m confused why you compute the loss using val data with respect to original “\theta”.)
2. In the “meta_update(*)” function, you use this (register_hook) to change the gradients. But the functions are(opt.zero_grad;  loss.backward) also used to compute the gradients. 
Do these functions(opt.zero_grad;  loss.backward) overwrite the previous gradients (changed by using the register_hook function)?
I’m looking forward to your reply.

**Comment by katerakelly** [comment]
1. We want to accumulate meta gradients across multiple tasks before using them for a meta-update, hence we accumulate a list of gradient dictionaries. Because the weights of `self.net` are not directly exposed, the simplest way to do the meta-update seemed to be registering a backward hook which overwrites the gradients computed in the "dummy" pass (L71, 86 of `maml.py`) with the accumulated meta gradients from the multiple tasks. 
2. The hook is not actually executed until `loss.backward()` is called (L86), so this is when the meta-grads are overwriting the dummy ones.
